pyface/data_view/data_models/tests/test_array_data_model.py

```diff
@@ -0,0 +1,335 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+from unittest import TestCase
+
+from traits.testing.unittest_tools import UnittestTools
+from traits.testing.optional_dependencies import numpy as np, requires_numpy
```

Comment:
For better or worse, Pyface depends on TraitsUI - this is likely not the only test that will fail if TraitsUI is not installed.

Maybe one day that will not be the case.

The problem here is that in a perfect world, importing traitsui should not force toolkit selection.

Reply:
I was actually hoping that it would be possible to test `ArrayDataModel` without requiring GUI toolkits. Is that really not possible?

Reply:
It is possible, but:

- if you don't have traitsui installed and are very careful about things, you might be able to run just these tests without selecting a toolkit (no guarantees, though)
- given that we have the code in traits to handle numpy dependency properly for tests, I don't want to duplicate that code in Pyface
- you'd hope that you could test _traits_ without requiring GUI toolkit selection, but we can't

The villain is traitsui.api, and that's what should be fixed (but that may take some effort).

Reply:
> given that we have the code in traits to handle numpy dependency properly for tests, I don't want to duplicate that code in Pyface

On this, my main concern was related to https://github.com/enthought/traits/issues/1213
The `optional_dependencies` is understandably written with traits being its sole consumer. The import on `traitsui.api` can be removed, and the `requires_numpy` function can be renamed, all without backward compatibility constraints. If pyface imports from it, pyface is on its own: it can't rely on traits keeping this module the way it is. Trivial changes in `traits.testing.optional_dependencies` will cause trivial breakages on pyface tests. This will cause headache for those whose job are building packages and working out package versions compatibility. Unless these test utilities are pulled out to another common package that both traits and pyface depend on (I don't think that's worth it btw), I think it is totally worth it to maintain these simple functions locally in pyface even if they are duplications, for the purpose of decoupling.

> if you don't have traitsui installed and are very careful about things, you might be able to run just these tests without selecting a toolkit (no guarantees, though)

We actually can't run these tests without loading a GUI toolkit early, because the callable for `dispatch="ui"` would otherwise be undefined and the trait change handler fails immediately. Maybe we can revisit this once we have audited all the `dispatch="ui"` here...

Reply:
> I think it is totally worth it to maintain these simple functions locally in pyface even if they are duplications, for the purpose of decoupling.

Just wanted to clarify that it is okay if this is not changed here, but I will probably come back to change this in a separate PR.

Reply:
I'd disagree on this.  Pyface depends strongly on Traits; and I really don't like duplication of code. I've run into too many situations where we fix a bug in one place only to find the same bug somewhere else because of a local copy of identical code.

The one thing that probably should happen is that these methods should be promoted to an `pyface.testing.api` module with things like `UnittestTool`.

Reply:
Should there be an issue on traits to move `requires_numpy` and `numpy` to `traits.testing.api`?

Reply:
Opened https://github.com/enthought/traits/issues/1236